File: src-csharp/Vla.Engine/TopologicalSorter.cs
using System.Collections.Immutable;

namespace Vla.Engine;

/// <summary>
///     A topological sorter for nodes.
/// </summary>
public readonly struct TopologicalSorter(params (string from, string to)[] connections)
{
	public ImmutableDictionary<string, ImmutableArray<string>> Graph { get; } =
		CreateGraph(connections.ToImmutableArray());

	private static ImmutableDictionary<string, ImmutableArray<string>> CreateGraph(
		ImmutableArray<(string from, string to)> connections)
	{
		var graph = new Dictionary<string, ImmutableArray<string>>();

		foreach (var connection in connections)
		{
			if (!graph.ContainsKey(connection.from))
				graph.Add(connection.from, ImmutableArray<string>.Empty);

			if (!graph.ContainsKey(connection.to))
				graph.Add(connection.to, ImmutableArray<string>.Empty);

			graph[connection.from] = graph[connection.from]
				.Add(connection.to)
				.Distinct()
				.ToImmutableArray();
		}

		return graph.ToImmutableDictionary();
	}

	public ImmutableArray<string> Sort()
	{
		var visited = new HashSet<string>();
		var stack = new Stack<string>();

		foreach (var nodeId in Graph.Keys)
			if (!visited.Contains(nodeId))
				Visit(nodeId, ref visited, ref stack);

		return stack.ToImmutableArray();
	}

	private void Visit(string nodeId, ref HashSet<string> visited, ref Stack<string> stack)
	{
		visited.Add(nodeId);

		foreach (var childId in Graph[nodeId])
			if (!visited.Contains(childId))
				Visit(childId, ref visited, ref stack);

		stack.Push(nodeId);
	}
}

File: src/core/engine/topological_sorter.ts
export function sort(values: string[][]): string[] {
    let stack: string[] = [];
    let visited: Record<string, boolean> = {};

    let graph: Record<string, string[]> = {};
    for (let [a, b] of values) {
        if (!graph[a]) graph[a] = [];
        graph[a].push(b);
    }

    function visit(node: string) {
        if (visited[node]) return;
        visited[node] = true;

        let neighbors = graph[node] || [];
        for (let neighbor of neighbors) {
            visit(neighbor);
        }

        stack.push(node);
    }

    for (let node in graph) {
        visit(node);
    }

    stack.reverse();

    return stack;
}
